File: codeforces/round379/C.cc
// Codeforces Round #379 (Div. 2)
// http://codeforces.com/contest/734

// ID: 734C (Anton and Making Potions)

#include <iostream>
#include <stdio.h>
#include <vector>
#include <algorithm>
using namespace std;

const int MAX = 2 * 1e5 + 1;
long long int A[MAX], B[MAX], C[MAX], D[MAX];

void solve() {
	long long int n; // num potions
	long long int m, k; // num spells of each type
	long long int x; // portion creation time
	long long int s; // manapoint usage limit
	cin >> n >> m >> k >> x >> s;

	for(int i=0; i<m; i++) scanf("%lld", A + i);
	for(int i=0; i<m; i++) scanf("%lld", B + i);
	for(int i=0; i<k; i++) scanf("%lld", C + i);
	for(int i=0; i<k; i++) scanf("%lld", D + i);

	// don't use first spell
	long long int best = 0;
	int freePotion = C[upper_bound(D, D+k, s) - D - 1];
	if(freePotion < n){
		best = (n - freePotion) * x;
	}

	// use first spell
	for(int i=0; i<m; i++){
		long long int cand = 0;
		long long int remain = n;
		long long int mana = s;
		if(B[i] > s) continue;
		// use first type spell
		mana -= B[i];
		// use second type spell
		int ub = upper_bound(D, D+k, mana) - D - 1;
		remain -= C[ub];
		if(remain > 0) cand += remain * A[i];
		best = min(best, cand);
	}

	cout << best << endl;
}

int main() {
	solve();
	return 0;
}

File: codeforces/round379/C.hs
-- Codeforces Round #379 (Div. 2)
-- http://codeforces.com/contest/734

-- ID: 734C (Anton and Making Potions)

import Control.Monad
import Data.Array
import Data.Int
import Data.Maybe
import qualified Data.ByteString as BS
import qualified Data.ByteString.Char8 as BS8

getInt64s = (map (fromIntegral . fst . fromJust . BS8.readInteger) . BS8.words) `fmap` BS.getLine :: IO [Int64]

main = do
	[n, m, k] <- getInt64s
	[x, s] <- getInt64s
	as <- (listArray (0, m-1)) `fmap` getInt64s
	bs <- (listArray (0, m-1)) `fmap` getInt64s
	cs <- (listArray (0, k-1)) `fmap` getInt64s
	ds <- (listArray (0, k-1)) `fmap` getInt64s
	let ub = upperBound ds k s
	let freePotion = cs ! (ub - 1)
	let noFirst = if ub > 0 then (max (n - freePotion) 0) * x else n * x
	let bestUsingFirst = minimum $ map (make as bs cs ds n k s) [0..m-1]
	print $ min noFirst bestUsingFirst

make as bs cs ds n k s i
	| bs ! i > s = maxBound :: Int64
	| otherwise = (max 0 remain) * (as ! i)
	where
		remain = n - (if ub == 0 then 0 else (cs ! (ub - 1)))
		ub = upperBound ds k (s - bs ! i)


upperBound :: Array Int64 Int64 -> Int64 -> Int64 -> Int64
upperBound ary n x = go 0 n where
	go lb ub
		| lb >= ub = ub
		| ary ! mid <= x = go (mid+1) ub
		| ary ! mid > x = go lb mid
		where mid = (lb + ub) `div` 2
